### PR TITLE
*: add docs for TiFlash MinTSO scheduler

### PR DESCRIPTION
**Pull request by windtalker**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v8.4 (TiDB 8.4 versions)
- [x] v8.3 (TiDB 8.3 versions)
- [x] v8.2 (TiDB 8.2 versions)
- [x] v8.1 (TiDB 8.1 versions)
- [x] v7.5 (TiDB 7.5 versions)
- [x] v7.1 (TiDB 7.1 versions)
- [x] v6.5 (TiDB 6.5 versions)
- [x] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions))

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s): https://github.com/pingcap/docs-cn/pull/18542

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: tiflash/tiflash-mintso-scheduler.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+为了避免系统陷入死锁的状态，我们引入了两层 thread 的限制：
+
+* threads_soft_limit
+* threads_hard_limit
+
+其中 soft limit 主要用来限制系统使用的线程数，但是对于特定的 MPPTask，为了避免死锁可以打破这个限制，而 hard limit 则是为了保护系统，一旦超过 hard limit，TiFlash 可以通过报错来避免系统陷入死锁状态。
+
+利用 soft limit 来避免死锁的思想很简单，系统中存在一个特殊的 query，该 query 的所有 MPPTask 在调度时都可以突破 soft limit 的限制，这样只要系统的 thread 不超过 hard limit，系统中就必定存在一个 query 的所有 MPPTask 都可以正常执行，这样系统就不会出现死锁。
+
+MinTSO Scheduler 的目标就是在控制系统线程数的同时，确保系统中始终有一个特殊的 query，其所有的 MPPTask 都可以被调度到。 MinTSO Scheduler 是一个完全分布式的调度器，每个 TiFlash 仅根据自身信息对 MPPTask 进行调度，这样每个 TiFlash 的 MinTSO Scheduler 需要找到同一个“特殊”的 query。在 TiDB 中，每个 query 都会带有一个读的时间戳(TiDB 中称之为 start_ts)，MinTSO Scheduler 定义“特殊” query 的标准即为当前 TiFlash 节点上 start_ts 最小的 query，根据全局最小一定是局部最小的原理，所有的 TiFlash 选出的“特殊” query 必然是同一个。我们称之为 MinTSO query。MinTSO scheduler 的调度流程如下：
```
Comment (joey-yez) [inline tiflash/tiflash-mintso-scheduler.md:55]:
```suggestion
MinTSO Scheduler 的目标就是在控制系统线程数的同时，确保系统中始终有且只有一个特殊的 query，其所有的 MPPTask 都可以被调度到。 MinTSO Scheduler 是一个完全分布式的调度器，每个 TiFlash 仅根据自身信息对 MPPTask 进行调度，因此，所有 TiFlash 的 MinTSO Scheduler 需要找到同一个“特殊”的 query。在 TiDB 中，每个 query 都会带有一个读的时间戳(TiDB 中称之为 start_ts)，MinTSO Scheduler 定义“特殊” query 的标准即为当前 TiFlash 节点上 start_ts 最小的 query，根据全局最小一定是局部最小的原理，所有的 TiFlash 选出的“特殊” query 必然是同一个。我们称之为 MinTSO query。MinTSO scheduler 的调度流程如下：
```

File: tiflash/tiflash-mintso-scheduler.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+上面的 query 会在每个 TiFlash 节点中生成 2 个 MPPTask，其中 `ExchangeSender_21` 所在的 MPPTask 依赖于 `ExchangeSender_45` 所在的 MPPTask。假设这个 query 高并发的情况下，调度器调度了每个 query 中 `ExchangeSender_45` 所在的 MPPTask，则系统就会进入死锁的状态。
+
+为了避免系统陷入死锁的状态，我们引入了两层 thread 的限制：
+
+* threads_soft_limit
+* threads_hard_limit
+
+其中 soft limit 主要用来限制系统使用的线程数，但是对于特定的 MPPTask，为了避免死锁可以打破这个限制，而 hard limit 则是为了保护系统，一旦超过 hard limit，TiFlash 可以通过报错来避免系统陷入死锁状态。
+
+利用 soft limit 来避免死锁的思想很简单，系统中存在一个特殊的 query，该 query 的所有 MPPTask 在调度时都可以突破 soft limit 的限制，这样只要系统的 thread 不超过 hard limit，系统中就必定存在一个 query 的所有 MPPTask 都可以正常执行，这样系统就不会出现死锁。
```
Comment (joey-yez) [inline tiflash/tiflash-mintso-scheduler.md:53]:
```suggestion
使用 soft limit 和 hard limit 避免死锁的思想是：通过 soft limit 限制所有 query 使用的线程资源总量，充分利用资源的基础上，避免线程资源耗尽；通过 hard limit 确保在任何情况下，系统中至少存在一个 query 可以突破 soft limit 的限制，继续获取线程资源并继续运行，避免死锁。只要线程数不超过 hard limit，系统中就必定存在一个 query 的所有 MPPTask 都可以正常执行，这样系统就不会出现死锁。
```

File: tiflash/tiflash-mintso-scheduler.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+使用 soft limit 和 hard limit 避免死锁的思想是：通过 soft limit 限制所有 query 使用的线程资源总量，充分利用资源的基础上，避免线程资源耗尽；通过 hard limit 确保在任何情况下，系统中至少存在一个 query 可以突破 soft limit 的限制，继续获取线程资源并继续运行，避免死锁。只要线程数不超过 hard limit，系统中就必定存在一个 query 的所有 MPPTask 都可以正常执行，这样系统就不会出现死锁。
+
+MinTSO Scheduler 的目标就是在控制系统线程数的同时，确保系统中始终有且只有一个特殊的 query，其所有的 MPPTask 都可以被调度到。 MinTSO Scheduler 是一个完全分布式的调度器，每个 TiFlash 仅根据自身信息对 MPPTask 进行调度，因此，所有 TiFlash 的 MinTSO Scheduler 需要找到同一个“特殊”的 query。在 TiDB 中，每个 query 都会带有一个读的时间戳(TiDB 中称之为 start_ts)，MinTSO Scheduler 定义“特殊” query 的标准即为当前 TiFlash 节点上 start_ts 最小的 query，根据全局最小一定是局部最小的原理，所有的 TiFlash 选出的“特殊” query 必然是同一个。我们称之为 MinTSO query。MinTSO scheduler 的调度流程如下：
+
+<img src="/media/tiflash/tiflash_mintso_v2.png" width=50%></img>
+
+通过引入 soft limit 与 hard limit，MinTSO 能保证系统线程数的同时有效地避免了系统死锁。不过对于高并发场景，可能会出现大部份 query 都只有部分 MPPTask 被调度的情况，对于只有部分 MPPTask 被调度的 query，它们实际上无法正常执行，从而导致系统执行效率低下。为了避免这种情况，我们给 MinTSO Scheduler 在 query 层面引入了一个限制，即 active_query_soft_limit，该限制的意思是系统最多只有 active_query_soft_limit 个 query 的 MPPTask 可以参与调度，对于其他的 query，其 MPPTask 不参与调度，只有等当前 query 结束之后，新的 query 才能参与调度。当然这个限制只是一个 soft limit，因为对于 MinTSO query 来说，其所有 MPPTask 在系统线程数不超过 hard limit 的时候都可以直接被调度。
```
Comment (lilin90) [inline tiflash/tiflash-mintso-scheduler.md:59]:
```suggestion
通过引入 soft limit 与 hard limit，MinTSO 调度器在控制系统线程数的同时，有效地避免了系统死锁。不过对于高并发场景，可能会出现大部份查询都只有部分 MPP Task 被调度的情况。只有部分 MPP Task 被调度的查询实际上无法正常执行，从而导致系统执行效率低下。为了避免这种情况，TiFlash 给 MinTSO Scheduler 在查询层面引入了一个限制，即 active_query_soft_limit，该限制的意思是系统最多只有 active_query_soft_limit 个查询的 MPP Task 可以参与调度；对于其它的查询，其 MPP Task 不参与调度，只有等当前查询结束之后，新的查询才能参与调度。当然该限制只是一个 soft limit，因为对于 MinTSO 查询来说，其所有 MPP Task 在系统线程数不超过 hard limit 时都可以直接被调度。
```

**Comment by lilin90** [comment]
@windtalker Please take a look at the comments.

File: TOC.md [hunk 4]
```diff
@@ -676,6 +676,7 @@
     - [TiFlash 延迟物化](/tiflash/tiflash-late-materialization.md)
     - [TiFlash 数据落盘](/tiflash/tiflash-spill-disk.md)
     - [TiFlash 数据校验](/tiflash/tiflash-data-validation.md)
+    - [TiFlash MinTSO 调度器](/tiflash/tiflash-mintso-scheduler.md)
```
[the remark this answers is not in the document]
Comment (windtalker) [reply TOC.md:679]: 这里应该可以

File: tiflash/tiflash-mintso-scheduler.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
+---
+title: TiFlash MinTSO Scheduler
+summary: 介绍 TiFlash MinTSO 调度器。
+---
+
+# TiFlash MinTSO Scheduler
```
[the remark this answers is not in the document]
Comment (windtalker) [reply tiflash/tiflash-mintso-scheduler.md:6]: 可以的

File: tiflash/tiflash-mintso-scheduler.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+在 TiDB 中，对于 MPP query，TiDB 会将 query 拆成一个或多个 MPPTask，并将 MPPTask 发送给对应的 TiFlash 节点。而在 TiFlash 节点内部，对每个 MPPTask 进行编译与执行。在 TiFlash 使用 [pipeline 执行模型](/tiflash/tiflash-pipeline-model.md) 之前，对于每个 MPPTask，TiFlash 都需要使用若干个线程（具体线程数取决于 MPPTask 的复杂度以及 TiFlash 并发参数的设置）来执行。在高并发场景中，TiFlash 节点会同时接受到多个 MPPTask，如果 TiFlash 不对 MPPTask 的执行有所控制，则 TiFlash 需要向系统申请的线程数会随着 MPPTask 数量的增加而线性增加。过多的线程一方面会影响 TiFlash 的执行效率，另一方面操作系统本身支持的线程数也是有限的，当 TiFlash 申请的线程数超过操作系统的限制时，TiFlash 就会碰到无法申请线程的错误。
+
+为了提升在高并发场景下 TiFlash 的处理能力，我们需要在 TiFlash 中引入一个 task 的 scheduler。
+
+## 设计实现
+
+如背景所述，TiFlash task scheudler 引入的初衷是控制运行时使用的线程数。一个简单的想法是指定 TiFlash 可以申请的最大线程数，对于每个 MPPTask，调度器根据当前系统已经使用的线程数以及该 MPPTask 预期使用的线程数，决定该 MPPTask 是否能够被调度：
+
+<img src="/media/tiflash/tiflash_mintso_v1.png" width=50%></img>
```
[the remark this answers is not in the document]
Comment (lilin90) [reply tiflash/tiflash-mintso-scheduler.md:20]: 使用这种格式的链接呈现图片有一个弊端：在网页上能按照指定的 width 显示，在文档站右上角支持下载的 PDF 里图片完全显示不出来。所以只要不是特别不协调，能避免就尽量避免。

**Review by lilin90 — APPROVED** [review]
/lgtm

**Review by joey-yez — APPROVED** [review]

**Review by qiancai — APPROVED** [review]
Rest LGTM

File: tiflash/tiflash-mintso-scheduler.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+TiFlash MinTSO 调度器是 TiFlash 中一个分布式的基于 MPP Task 的调度器。本文介绍 TiFlash MinTSO 调度器的实现原理。
+
+## 背景
+
+对于 MPP 查询，TiDB 会将查询拆成一个或多个 MPP Task，并将 MPP Task 发送给对应的 TiFlash 节点。在 TiFlash 节点内部，对每个 MPP Task 进行编译与执行。在 TiFlash 使用 [pipeline 执行模型](/tiflash/tiflash-pipeline-model.md)之前，对于每个 MPP Task，TiFlash 都需要使用若干个线程来执行，具体线程数取决于 MPP Task 的复杂度以及 TiFlash 并发参数的设置。
+
+在高并发场景中，TiFlash 节点会同时接收到多个 MPP Task，如果 TiFlash 不对 MPP Task 的执行有所控制，则 TiFlash 需要向系统申请的线程数会随着 MPP Task 数量的增加而线性增加。过多的线程一方面会影响 TiFlash 的执行效率，另一方面操作系统本身支持的线程数有限，当 TiFlash 申请的线程数超过操作系统的限制时，TiFlash 就会遇到无法申请线程的错误。
+
+为了提升 TiFlash 在高并发场景下的处理能力，需要在 TiFlash 中引入一个 Task 调度器。
```
Comment (qiancai) [inline tiflash/tiflash-mintso-scheduler.md:16]:
```suggestion
为了提升 TiFlash 在高并发场景下的处理能力，需要在 TiFlash 中引入一个 MPP Task 调度器。
```

**Review by windtalker — COMMENTED** [review]
update

File: tiflash/tiflash-mintso-scheduler.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+|           │ └─ExchangeSender_21            | 9990.00  | mpp[tiflash] |               | ExchangeType: Broadcast, Compression: FAST               |
+|           │   └─Selection_20               | 9990.00  | mpp[tiflash] |               | not(isnull(test.t0.id))                                  |
+|           │     └─TableFullScan_19         | 10000.00 | mpp[tiflash] | table:a       | pushed down filter:empty, keep order:false, stats:pseudo |
+|           └─Selection_24(Probe)            | 9990.00  | mpp[tiflash] |               | not(isnull(test.t0.id))                                  |
+|             └─TableFullScan_23             | 10000.00 | mpp[tiflash] | table:b       | pushed down filter:empty, keep order:false, stats:pseudo |
++--------------------------------------------+----------+--------------+---------------+----------------------------------------------------------+
+```
+
+例如，以上查询会在每个 TiFlash 节点中生成 2 个 MPP Task，其中 `ExchangeSender_45` 所在的 MPP Task 依赖于 `ExchangeSender_21` 所在的 MPP Task。在该查询高并发的情况下，如果调度器调度了每个查询中 `ExchangeSender_45` 所在的 MPP Task，系统就会进入死锁状态。
```
Comment (lilin90) [inline tiflash/tiflash-mintso-scheduler.md:49]:
```suggestion
例如，以上查询会在每个 TiFlash 节点中生成 2 个 MPP Task，其中执行器 `ExchangeSender_45` 所在的 MPP Task 依赖于 `ExchangeSender_21` 所在的 MPP Task。在该查询高并发的情况下，如果调度器调度了每个查询中 `ExchangeSender_45` 所在的 MPP Task，系统就会进入死锁状态。
```

**Pull request merged**